File: 09-Linked-List/01-Singly-Linked-List/cpp/reverse_linked_list.cpp
/*
 * Problem: Reverse Linked List
 * Source:  LeetCode 206
 * Time: O(n), Space: O(1)
 * Tags: linked-list
 */
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr, *curr = head;
    while (curr) {
        ListNode* next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    
    ListNode* reversed = reverseList(head);
    while (reversed) { cout << reversed->val << " "; reversed = reversed->next; }
    // Output: 5 4 3 2 1
    return 0;
}

File: 09-Linked-List/01-Singly-Linked-List/java/ReverseLinkedList.java
public class ReverseLinkedList {
    static class ListNode { int val; ListNode next; ListNode(int x) { val = x; } }
    public static ListNode reverseList(ListNode head) {
        ListNode prev = null, curr = head;
        while (curr != null) { ListNode next = curr.next; curr.next = prev; prev = curr; curr = next; }
        return prev;
    }
    public static void main(String[] args) {
        ListNode head = new ListNode(1); head.next = new ListNode(2); head.next.next = new ListNode(3);
        ListNode rev = reverseList(head);
        while (rev != null) { System.out.print(rev.val + " "); rev = rev.next; } // 3 2 1
    }
}
